packages/turndown/src/commonmark-rules.js

```diff
@@ -478,6 +492,23 @@ rules.code = {
   }
 }
 
+// Fix: Web clipper has trouble with code blocks on Joplin's website.
+// See https://github.com/laurent22/joplin/pull/10126#issuecomment-2016523281 .
+// Web clipper clippering contents is all in oneline. The format features are: <pre ...><code ...><span class="token-line">. Span with class of "token-line" represent one line. 
+// Test case: packages/app-cli/tests/html_to_md/code_multiline_3.html .
+rules.joplinOrgTokenLineSpan = {
+  filter: function (node) {
+    const grandfather = node.parentNode.parentNode?? null;
+    return node.nodeName === 'SPAN' && node.getAttribute('class') === 'token-line' && grandfather && isCodeBlock(grandfather); 
+  },
+
+  replacement: function (content, node, options) {
+    content = content.replace(/\r?\n|\r/g, '');
+    // If content replaced '\r\n' is empty, it indicates that the line is empty line; if not, keep the leading before and after content.
+    return content === '' ? '\n\n' :  node.flankingWhitespace.leading + content + node.flankingWhitespace.trailing + '\n';
+  }
+}
+
```

Comment:
I'm struggling to review this, I don't really understand what it does

Reply:
Web clipper clipping contents from Joplin's website is all in one line. The format features are: <pre ...><code ...><span class="token-line"><span class="...">....  Span with class of "token-line" represent one line.

Reply:
At present, users can write plugins that use Renderer's plugin API to affect the process of MdToHtml. At the same time, users may also need Turndown's plugin API to write plugins that affect the process of HtmlToMd. Please consider adding a plugin API for Turndown.

Reply:
> @personalizedrefrigerator, was this for the bug you mentioned, and does the fix make sense to you?

This was the bug I mentioned, however I'm not sure that the fix makes sense to me. I think the issue here is that the `<br/>`s at the end of Docusaurus code blocks aren't being converted to linebreaks by Turndown.

For context, I've attached simplified HTML for a Docusaurus code block (taken from [here](https://docusaurus.io/docs/create-doc#doc-front-matter)):
````html
<code class="codeBlockLines_FJaf">
    <span class="token-line" style="color: rgb(212, 212, 212);">
        <span class="token front-matter-block punctuation" style="color: rgb(212, 212, 212);">---</span>
        <br>
    </span>
    <span class="token-line" style="color: rgb(212, 212, 212);">
        <span class="token front-matter-block"></span>
        <span class="token front-matter-block front-matter yaml language-yaml key atrule" style="color: rgb(86, 156, 214);">id</span>
        <span class="token front-matter-block front-matter yaml language-yaml punctuation" style="color: rgb(212, 212, 212);">:</span>
        [...]
        <br>
    </span>
    [...]
    <span class="token-line" style="color: rgb(212, 212, 212);">
        <span class="token front-matter-block"></span>
        <span class="token front-matter-block punctuation" style="color: rgb(212, 212, 212);">---</span>
        <br>
    </span>
</code> 
```
Observe that each (inline) `token-line` span ends with a `<br>` that previously wasn't detected by Turndown.

Reply:
Actually `<br/>` has been correctly parsed.
The key to the problem is, if span with class "token keyword", turndown added line breaks before and after span, please see the parsing results of `async` and `function`.

`nStart: async function() {` in code blocks, it html is:

```
<span class="token-line" style="color: rgb(57, 58, 52); --darkreader-inline-color: #c5c0b9;" data-darkreader-inline-color="">
	<span class="token plain">    </span>
	<span class="token function-variable function" style="color: rgb(215, 58, 73); --darkreader-inline-color: #da4b58;" data-darkreader-inline-color="">onStart</span>
	<span class="token operator" style="color: rgb(57, 58, 52); --darkreader-inline-color: #c5c0b9;" data-darkreader-inline-color="">:</span>
	<span class="token plain"> </span>
	<span class="token keyword" style="color: rgb(0, 0, 159); --darkreader-inline-color: #6da2ff;" data-darkreader-inline-color="">async</span>
	<span class="token plain"> </span>
	<span class="token keyword" style="color: rgb(0, 0, 159); --darkreader-inline-color: #6da2ff;" data-darkreader-inline-color="">function</span>
	<span class="token punctuation" style="color: rgb(57, 58, 52); --darkreader-inline-color: #c5c0b9;" data-darkreader-inline-color="">(</span>
	<span class="token punctuation" style="color: rgb(57, 58, 52); --darkreader-inline-color: #c5c0b9;" data-darkreader-inline-color="">)</span>
	<span class="token plain"> </span>
	<span class="token punctuation" style="color: rgb(57, 58, 52); --darkreader-inline-color: #c5c0b9;" data-darkreader-inline-color="">{</span>
	<span class="token plain"></span>
	<br>
</span>
```





![image](https://github.com/laurent22/joplin/assets/88243938/22de05f2-926c-4828-bb8c-687bf270e372)

Reply:
If this fix `packages/turndown/src/commonmark-rules.js/rules.joplinOrgTokenLineSpan` is not suitable, can I roll back this submission and only fix # 5626?

Reply:
Remove `packages/turndown/src/commonmark-rules.js/rules.joplinOrgTokenLineSpan`. Analyze the current code, identify the reasons for incorrect formatting in the code block, and preserve the link break formatting for the code block.